### PR TITLE
Feature/mask NaNs in training loss function

### PR DESCRIPTION
**Pull request by sahahner**
Variables with missing values that are imputed by the imputer should not be considered in the loss.

Solves issue [271](https://github.com/ecmwf-lab/aifs-mono/issues/271)

**Describe the solution**
Pass the product of the imputer NaN mask(s) to the loss in the first forward pass and multiply the contributions to the loss of imputed grid points by zero.
Also, apply the remapper to the mask to remap the NaN mask in case the remapper is used.

The NaN masks are prepared in the imputer. The remapper contains a new function to remap the NaN mask.
These changes are part of anemoi-model, PR [#56](https://github.com/ecmwf/anemoi-models/pull/56)

**Attention**
This changes the default behaviour when using variables that contain NaN values that are imputed.

### CODE REVIEW

**Comment by FussyDuck** [comment]
[CLA assistant check](https://bol-claassistant.ecmwf.int:443/ecmwf/anemoi-training?pullRequest=72) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://bol-claassistant.ecmwf.int:443/ecmwf/anemoi-training?pullRequest=72) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: sahahner<br/>:x: jakob-schloer<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://bol-claassistant.ecmwf.int:443/check/ecmwf/anemoi-training?pullRequest=72) it.</sub>

File: CHANGELOG.md
Comment (HCookie) [inline CHANGELOG.md:29]: I believe the wrong issue has been referenced, should this not be #72?

**Comment by HCookie** [comment]
The loss function code is being altered in #70 to enable flexible configuration of loss functions.
This PR will require some changes to be compatible.

## Changes 
- masks will need to be added to ‘mae’, and ‘logcosh’
- Initialisation of loss functions needs to be pulled from the other PR
- Setting of the mask will require a ‘hasattr’ check to determine if the loss function exposes that functionality.

File: CHANGELOG.md
Comment (jakob-schloer) [reply CHANGELOG.md:29]: There is both one in anemoi-training as well as in aifs-mono.

Comment (sahahner) [reply CHANGELOG.md:29]: Since there has been some discussion in aifs-mono, I want to have it linked. I listed both now.

**Comment by sahahner** [comment]
This has to be merged after #137

File: src/anemoi/training/train/forecaster.py
Comment (HCookie) [inline src/anemoi/training/train/forecaster.py:386]: What if a user has set the number of sanity_steps to 0, `validation` will not run before training, and thus the mask won't be set

File: src/anemoi/training/train/forecaster.py
Comment (HCookie) [inline src/anemoi/training/train/forecaster.py:237]: Would it not make sense to have this if at the start so that the above multiplications won't unneccesarily take place.

File: src/anemoi/training/train/forecaster.py
Comment (sahahner) [reply src/anemoi/training/train/forecaster.py:386]: The mask is set in the first application of the preprocessors, which are applied in both training and validation mode.

Comment (HCookie) [reply src/anemoi/training/train/forecaster.py:386]: But the scalar won't be set.

Comment (sahahner) [reply src/anemoi/training/train/forecaster.py:386]: Isn't the scalar of the loss function set in the initialization of the model?

Comment (HCookie) [reply src/anemoi/training/train/forecaster.py:386]: Sorry, the scalar will be set in init to `1`.
But, if the validation_step is not run before training, the scalar will not be updated with the true mask value.

File: src/anemoi/training/train/forecaster.py
Comment (HCookie) [reply src/anemoi/training/train/forecaster.py:228]: @sahahner Will this be addressed?

**Review by HCookie — APPROVED** [review]

**Review by jakob-schloer — APPROVED** [review]
Approve after testing.

**Pull request merged**
